### ISSUE TITLE
Local functions: squiggles all over the place when using an unsupported modifier.

### ISSUE THREAD

**Issue by balajikris** [dotnet/roslyn#8645]
The screenshot below shows the difference of **_adding/removing an unsupported `internal` modifier to a local function_**.
Test code included at the bottom of this report.

Reporting **two issues** here:
1. None of the 15 errors reported _point me to the actual cause_ - that local functions cannot have the `internal` modifier.
2. Need _better error recovery_ in the parser here - notice that the parser thinks all following top level methods are declared inside a namespace.

![image](https://cloud.githubusercontent.com/assets/4213867/13019450/68e04886-d184-11e5-8c08-8316f7823c69.png)

```C#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApplication6
{
    class Class1
    {

        public int GetF()
        {
            internal int GetF()
            {
                return 1;
            }

            return GetF();
        }

        public int Foo(int x)
        {
            Func<int, int> Foo = zzz => zzz + 1;

            // F(1);

            int LocalFunc(int y) => Foo(y) + 1;


            void F2(int y = 0, string str = "")
            {
                Console.WriteLine(1);
                Console.WriteLine(str);

            }

            F2(y: 1, str: null);

            dynamic v = 2;

            int GetVal()
            {
                return 1;
            }

            int z = GetVal();

            return LocalFunc(1);
        }

        int GetVal()
        {
            return 1;
        }

        static void Main(string[] args, int x)
        {
            void Local(int z, int y)
            {
            }

            Local(1, 2);
        }
    }
}
```

**Comment by khyperia** [comment]
Good to see local functions are getting a beating :smile:

This particular issue is more complicated than I originally thought, I ran into it last summer, so I figured I'd share my knowledge. The old parser's behavior (without local functions) does something like this, if I recall correctly:

1. Parsing method body...
2. Encounters `internal` token
3. Assumes the user forgot the closing brace on a method
4. Drops back to class-level declaration parsing
5. Re-parses the `internal` token as another method
6. Encounters the *actual* closing brace of the method, and is parsed as the closing brace of the class (causing the rest of the methods in the class to have lots of errors)

I left this behavior in the prototype, as I didn't see a clear easy solution. I did make a prototype that accepted the `internal` token as an error and continued like it was a local function, but this caused the unfortunate result that the rest of the methods in a class were parsed as local functions if the user *actually did* forget the closing brace of a method.

**Comment by controlflow** [comment]
What about missing closing braces?

```c#
class C {
  public void M() {
  public void G() { }
}

I was thinkinking Roslyn do not parses all the possible C# modifiers for local methods exactly to improve parser recovery in the case of missing closing brace from method above..
